windows/CodePush.Net46/UpdateManager.cs

```diff
@@ -0,0 +1,326 @@
+﻿using Newtonsoft.Json;
+using Newtonsoft.Json.Linq;
+using System.IO;
+using System.IO.Compression;
+using System.Threading.Tasks;
+using System.Runtime.CompilerServices;
+using System;
+#if WINDOWS_UWP
+using Windows.Web.Http;
+#else
+using CodePush.Net46.Adapters.Http;
+using PCLStorage;
+#endif
+
+[assembly: InternalsVisibleTo("CodePush.Net46.UnitTest")]
+
+namespace CodePush.ReactNative
+{
+    internal class UpdateManager
+    {
+        #region Internal methods
+
+        internal async Task ClearUpdatesAsync()
+        {
+            await (await GetCodePushFolderAsync().ConfigureAwait(false)).DeleteAsync().ConfigureAwait(false);
+        }
+
+        internal async Task DownloadPackageAsync(JObject updatePackage, string expectedBundleFileName, Progress<HttpProgress> downloadProgress)
+        {
+            // Using its hash, get the folder where the new update will be saved 
+            var codePushFolder = await GetCodePushFolderAsync().ConfigureAwait(false);
+            var newUpdateHash = (string)updatePackage[CodePushConstants.PackageHashKey];
+            var newUpdateFolder = await GetPackageFolderAsync(newUpdateHash, false).ConfigureAwait(false);
+            if (newUpdateFolder != null)
+            {
+                // This removes any stale data in newPackageFolderPath that could have been left
+                // uncleared due to a crash or error during the download or install process.
+                await newUpdateFolder.DeleteAsync().ConfigureAwait(false);
+            }
+
+            newUpdateFolder = await GetPackageFolderAsync(newUpdateHash, true).ConfigureAwait(false);
+            var newUpdateMetadataFile = await newUpdateFolder.CreateFileAsync(CodePushConstants.PackageFileName, CreationCollisionOption.ReplaceExisting).ConfigureAwait(false);
+            var downloadUrlString = (string)updatePackage[CodePushConstants.DownloadUrlKey];
+            var downloadFile = await GetDownloadFileAsync().ConfigureAwait(false);
+
+            await UpdateUtils.DownloadBundleAsync(downloadUrlString, downloadFile.Path, downloadProgress);
+
+            try
+            {
+                // Unzip the downloaded file and then delete the zip
+                var unzippedFolder = await GetUnzippedFolderAsync().ConfigureAwait(false);
+                /**
+                 * TODO:
+                 *  1) ZipFile.ExtractToDirectory is nor reliable and throws exception if:
+                 *      - folder exists already
+                 *      - path is too long (> 250 chars)
+                 *  
+                 *  2) Un-zipping is quite long operation. Does it make sense for async?
```

Comment:
>) Un-zipping is quite long operation [](start = 21, length = 36)

It depends on where this is originally called from.  I would assume that this work is already being performed on a background thread, so there would be no reason to create yet another task to continue this work.